{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eef3907-8262-4f0b-9f29-649047e44711",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import scipy as sp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sc.settings.set_figure_params(figsize=(5,5))\n",
    "sns.set(font_scale=2)\n",
    "fig_dir=\"../results/figures\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "733f1c84-b4ea-4a8e-8145-f58ef85ecad7",
   "metadata": {},
   "source": [
    "## Load raw data\n",
    "The barcode cutoff (kneeplot) from cellranger was not looking right for all samples. Therefore, load raw data and make custom filter thresholds per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "629f5246-8ba3-4264-9b7a-280e4728b655",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get metadata from samplesheet\n",
    "meta = pd.read_csv('../tables/samplesheet.csv',\n",
    "                   index_col='internal_id')\n",
    "meta.drop(axis='columns', labels=['fastq_1', 'fastq_2'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f970882e-f9d7-439b-b6df-408c72a1ff5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load cellranger .h5 feature matrices\n",
    "path_dir='/data/projects/2021/Grabherr-scRNAseq-mouse/30_nfcore_scrnaseq_v2-0-0/cellranger'\n",
    "adatas_list = []\n",
    "cnvan_key_l = []\n",
    "for ind, sample in zip(meta.index, meta.to_dict(orient=\"records\")):\n",
    "\n",
    "    path_h5ads = f'{path_dir}/sample-{ind}/outs/raw_feature_bc_matrix.h5'\n",
    "    tmp_adata = sc.read_10x_h5(path_h5ads)\n",
    "    \n",
    "    # save gene conversion key and switch index to ensembl ids before making unique\n",
    "    cnvan_key_l.append(tmp_adata.var.copy())\n",
    "    tmp_adata.var = tmp_adata.var.drop(columns=['feature_types','genome'])\n",
    "    tmp_adata.var_names_make_unique()\n",
    "    assert tmp_adata.obs_names.is_unique\n",
    "    tmp_adata.obs = tmp_adata.obs.assign(**sample)\n",
    "    adatas_list.append(tmp_adata)\n",
    "\n",
    "for k in cnvan_key_l[1:]:\n",
    "    assert np.all(k==cnvan_key_l[0])\n",
    "\n",
    "cnvan_key = cnvan_key_l[-1]\n",
    "adata_raw = ad.concat(adatas_list, index_unique=\"_\")\n",
    "\n",
    "# Use conversion key to re-assign symbols to ensembl ids\n",
    "adata_raw.var.loc[cnvan_key.index,'gene_ids'] = cnvan_key.gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fc0756b-03df-48d3-8bdf-0ec134574d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_raw.obs[\"sample\"] = pd.Categorical(adata_raw.obs[\"sample\"])\n",
    "adata_raw.obs[\"group\"] = pd.Categorical(adata_raw.obs[\"group\"])\n",
    "adata_raw.obs[\"sex\"] = pd.Categorical(adata_raw.obs[\"sex\"])\n",
    "adata_raw.obs[\"batch\"] = pd.Categorical(adata_raw.obs[\"batch\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e6049e-cdf4-4ee2-b55b-7864372cd567",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Requires \"adata_raw.obs['sample']\" to be categorical\n",
    "sample_d = dict()\n",
    "for s in adata_raw.obs['sample'].values.unique():\n",
    "    _sampli = adata_raw[adata_raw.obs[\"sample\"] == s, :]\n",
    "    _sampli.obs[\"value\"] = 0\n",
    "    sample_d[s] = _sampli"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b411547-e579-452e-b5cc-1c38267c0244",
   "metadata": {},
   "source": [
    "### Basic filtering + first look at per Sample metrics\n",
    "Apply most basic filter thresholds per sample. Mostly removes empty barcodes without counts or genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6410bd14-5f61-410e-9fe8-b7c9d481691d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in adata_raw.obs['sample'].values.unique():\n",
    "    print(s + \":\")\n",
    "    print(sample_d[s].shape)\n",
    "    \n",
    "    fig, ax = plt.subplots(1,1,figsize=(10,10))\n",
    "    sc.pl.highest_expr_genes(sample_d[s], n_top=20, ax=ax)\n",
    "    \n",
    "    sc.pp.filter_cells(sample_d[s], min_genes=30)\n",
    "    sc.pp.filter_cells(sample_d[s], min_counts=30)\n",
    "    # Calculate QC metrics for all samples\n",
    "    sample_d[s].var['mito'] = sample_d[s].var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    sc.pp.calculate_qc_metrics(sample_d[s], qc_vars=['mito'], percent_top=None, log1p=False, inplace=True)\n",
    "    \n",
    "    #sample_d[s] = sample_d[s][sample_d[s].obs.pct_counts_mito < 40, :]\n",
    "    \n",
    "    print(\"After min filter:\")\n",
    "    print(sample_d[s].shape)\n",
    "    \n",
    "    fig, ax = plt.subplots(1,1,figsize=(10,10))\n",
    "    sc.pl.highest_expr_genes(sample_d[s], n_top=20, ax=ax)\n",
    "    \n",
    "    # Plot QC metrics\n",
    "    sc.pl.violin(sample_d[s], ['n_genes_by_counts', 'total_counts', 'pct_counts_mito'],\n",
    "             jitter=0.4, multi_panel=True)\n",
    "\n",
    "# After filtering individual samples concat to make one adata object\n",
    "adatas_new_l = []\n",
    "for s in adata_raw.obs['sample'].values.unique():\n",
    "    adatas_new_l.append(sample_d[s])\n",
    "adata_sub = ad.concat(adatas_new_l)\n",
    "\n",
    "adata = adata_raw[adata_sub.obs.index, adata_sub.var.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c867451-c61e-47de-97de-77d567b1f568",
   "metadata": {},
   "source": [
    "### Remove outliers\n",
    "Within each sample look for outliers (mito, ribo, hemoglobin, and housekeeping genes). Last one should hopefully get rid of doublets.\n",
    "For mito, additional global cutoff > 25%, to get a better colour scale in the scatter plots. I guess everything above 25 is not viable ...\n",
    "Mito cutoff will be lower for downstream filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5c9c02-bbc0-41b2-87f3-574a23401e38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter outliers\n",
    "def is_outlier(adata, metric: str, nmads: int):\n",
    "    M = adata.obs[metric]\n",
    "    outlier = (M < np.median(M) - nmads * M.mad()) | (\n",
    "        np.median(M) + nmads * M.mad() < M\n",
    "    )\n",
    "    return outlier\n",
    "\n",
    "sample_d = dict()\n",
    "for s in adata.obs['sample'].values.unique():\n",
    "    print(s)\n",
    "    adata_f = adata[adata.obs[\"sample\"] == s, :].copy()\n",
    "    adata_f.obs[\"value\"] = 0\n",
    "\n",
    "    # mitochondrial genes\n",
    "    adata_f.var[\"mito\"] = adata_f.var_names.str.startswith(\"mt-\")\n",
    "    # ribosomal genes\n",
    "    adata_f.var[\"ribo\"] = adata_f.var_names.str.startswith((\"Rps\", \"Rpl\", \"Gm42418\"))\n",
    "    # hemoglobin genes.\n",
    "    adata_f.var[\"hb\"] = adata_f.var_names.str.contains((\"Hb\"))\n",
    "    # housekeeping genes\n",
    "    adata_f.var[\"house\"] = adata_f.var_names.str.endswith((\"Actb\", \"Atp5f1\" , \"B2m\", \"Gapdh\", \"Hprt\", \"Pgk1\", \"Rer1\", \"Rpl13a\", \"Rpl27\", \"Sdha\", \"Tbp\", \"Ubc\"))\n",
    "\n",
    "    sc.pp.calculate_qc_metrics(\n",
    "        adata_f, qc_vars=[\"mito\", \"ribo\", \"hb\", \"house\"], inplace=True, percent_top=[20], log1p=True\n",
    "    )\n",
    "\n",
    "    adata_f.obs[\"outlier\"] = (\n",
    "        is_outlier(adata_f, \"log1p_total_counts\", 5)\n",
    "        | is_outlier(adata_f, \"log1p_n_genes_by_counts\", 5)\n",
    "        | is_outlier(adata_f, \"pct_counts_in_top_20_genes\", 5)\n",
    "    )\n",
    "    print(\"outlier\")\n",
    "    print(adata_f.obs.outlier.value_counts())\n",
    "\n",
    "    adata_f.obs[\"mito_outlier\"] = is_outlier(adata_f, \"pct_counts_mito\", 4) | (\n",
    "        adata_f.obs[\"pct_counts_mito\"] > 25\n",
    "    )\n",
    "    print(\"mito_outlier & pct_counts_mito > 25%\")\n",
    "    print(adata_f.obs.mito_outlier.value_counts())\n",
    "\n",
    "    adata_f = adata_f[(~adata_f.obs.outlier) & (~adata_f.obs.mito_outlier)].copy()\n",
    "    sample_d[s] = adata_f\n",
    "\n",
    "# After filtering individual samples concat to make one adata object\n",
    "adatas_new_l = []\n",
    "for s in adata.obs['sample'].values.unique():\n",
    "    adatas_new_l.append(sample_d[s])\n",
    "adata_sub = ad.concat(adatas_new_l)\n",
    "\n",
    "adata_outlier = adata[adata_sub.obs.index, adata_sub.var.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e21580a-c796-431a-95e0-3b63223fae19",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_d = dict()\n",
    "for s in adata_outlier.obs['sample'].values.unique():\n",
    "    _sampli = adata_outlier[adata_outlier.obs[\"sample\"] == s, :]\n",
    "    _sampli.obs[\"value\"] = 0\n",
    "    sample_d[s] = _sampli"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63a81495-9940-4f9b-83f6-23826cc925bd",
   "metadata": {},
   "source": [
    "### Second look at per Sample metrics after outlier filtering\n",
    "mt genes should be gone from top 20 highest_expr_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3ddddf5-bc14-4a65-8687-5b021d0759b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in adata_outlier.obs['sample'].values.unique():\n",
    "    print(s + \":\")\n",
    "    print(sample_d[s].shape)\n",
    "    \n",
    "    sc.pp.filter_genes(sample_d[s], min_cells=3)\n",
    "\n",
    "    # Calculate QC metrics for all samples\n",
    "    sample_d[s].var['mito'] = sample_d[s].var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    sc.pp.calculate_qc_metrics(sample_d[s], qc_vars=['mito'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "    print(\"After min filter:\")\n",
    "    print(sample_d[s].shape)\n",
    "    \n",
    "    fig, ax = plt.subplots(1,1,figsize=(10,10))\n",
    "    sc.pl.highest_expr_genes(sample_d[s], n_top=20, ax=ax)\n",
    "    \n",
    "    # Plot QC metrics\n",
    "    sc.pl.violin(sample_d[s], ['n_genes_by_counts', 'total_counts', 'pct_counts_mito'],\n",
    "             jitter=0.4, multi_panel=True)\n",
    "\n",
    "# After filtering individual samples concat to make one adata object\n",
    "adatas_new_l = []\n",
    "for s in adata_outlier.obs['sample'].values.unique():\n",
    "    adatas_new_l.append(sample_d[s])\n",
    "adata_sub = ad.concat(adatas_new_l)\n",
    "\n",
    "adata_filtered = adata_outlier[adata_sub.obs.index, adata_sub.var.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce22530-8d99-4f1d-a43a-70d1a6ee85eb",
   "metadata": {},
   "source": [
    "### Define per sample thresholds\n",
    "Use last code cell to visualize individual sample metrics and decide on cutoff. Still highly subjective cutoffs!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72252286-c2e7-4ab1-a231-be29036627b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FG-1\n",
    "dat = sample_d['FG-1']\n",
    "print(\"Sample: FG-1\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 300)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "# Use solo doublet detection instead!\n",
    "#sc.pp.filter_cells(dat, max_counts = 40000)\n",
    "#print('Number of cells after max count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 5, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 200)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-1'] = dat\n",
    "\n",
    "# FG-2\n",
    "dat = sample_d['FG-2']\n",
    "print(\"Sample: FG-2\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 150)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 10, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 100)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-2'] = dat\n",
    "\n",
    "# FG-3\n",
    "dat = sample_d['FG-3']\n",
    "print(\"Sample: FG-3\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 300)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 5, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 200)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-3'] = dat\n",
    "\n",
    "# FG-4\n",
    "dat = sample_d['FG-4']\n",
    "print(\"Sample: FG-4\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 300)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 5, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 200)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-4'] = dat\n",
    "\n",
    "# FG-5\n",
    "dat = sample_d['FG-5']\n",
    "print(\"Sample: FG-5\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 400)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 5, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 250)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-5'] = dat\n",
    "\n",
    "# FG-6\n",
    "dat = sample_d['FG-6']\n",
    "print(\"Sample: FG-6\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 300)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 10, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 250)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-6'] = dat\n",
    "\n",
    "# FG-7\n",
    "dat = sample_d['FG-7']\n",
    "print(\"Sample: FG-7\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 300)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 5, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 200)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "print(\" \")\n",
    "sample_d['FG-7'] = dat\n",
    "\n",
    "# FG-8\n",
    "dat = sample_d['FG-8']\n",
    "print(\"Sample: FG-8\")\n",
    "# Filter cells according to identified QC thresholds:\n",
    "print('Total number of cells: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_counts = 250)\n",
    "print('Number of cells after min count filter: {:d}'.format(dat.n_obs))\n",
    "dat = dat[dat.obs.pct_counts_mito < 5, :]\n",
    "print('Number of cells after MT filter: {:d}'.format(dat.n_obs))\n",
    "sc.pp.filter_cells(dat, min_genes = 200)\n",
    "print('Number of cells after gene filter: {:d}'.format(dat.n_obs))\n",
    "sample_d['FG-8'] = dat\n",
    "\n",
    "# After filtering individual samples concat to make one adata object\n",
    "adatas_new_l = []\n",
    "for s in adata_filtered.obs['sample'].values.unique():\n",
    "    adatas_new_l.append(sample_d[s])\n",
    "adata_sub = ad.concat(adatas_new_l)\n",
    "\n",
    "adata_filtered_per_sample = adata_filtered[adata_sub.obs.index, adata_sub.var.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de0a3e81-b8db-42c8-9042-6309216c6cf4",
   "metadata": {},
   "source": [
    "### Plot violin per Sample quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49c59468-287b-4b79-b6f9-396745afcfa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate QC metrics for all samples\n",
    "adata_filtered_per_sample.var['mito'] = adata_filtered_per_sample.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(adata_filtered_per_sample, qc_vars=['mito'], percent_top=None, log1p=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d89a6217-6e8b-40f5-862c-9d01dc2c021e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample quality plots\n",
    "fig, ax = plt.subplots(1,1,figsize=(30,10))\n",
    "sc.pl.violin(adata_filtered_per_sample, 'total_counts', groupby='sample', size=2, log=True, cut=0, ax=ax)\n",
    "\n",
    "fig, ax = plt.subplots(1,1,figsize=(30,10))\n",
    "sc.pl.violin(adata_filtered_per_sample, 'pct_counts_mito', groupby='sample', ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c15a21fb-428e-4778-b843-e8332dde38e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1,1,figsize=(10,10))\n",
    "sc.pl.highest_expr_genes(adata_filtered_per_sample, n_top=20, ax=ax)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f6ccab7-0b68-4e3c-a002-221041315c80",
   "metadata": {},
   "source": [
    "Usefull plots to decide on thresholds:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4f2d7e0-3ed4-4822-b639-4c77c55ea807",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Decide on Filter thresholds\n",
    "dat = adata_filtered_per_sample\n",
    "\n",
    "p1 = sc.pl.scatter(dat, 'total_counts', 'n_genes_by_counts', color='pct_counts_mito')\n",
    "p2 = sc.pl.scatter(dat[dat.obs['total_counts']<5000], 'total_counts', 'n_genes_by_counts', color='pct_counts_mito')\n",
    "\n",
    "sc.pl.violin(dat, ['n_genes_by_counts', 'total_counts', 'pct_counts_mito'],\n",
    "             jitter=0.4, multi_panel=True)\n",
    "\n",
    "#Thresholding decision: counts\n",
    "p3 = sns.displot(dat.obs['total_counts'], kde=False, bins=60)\n",
    "plt.show()\n",
    "\n",
    "p4 = sns.displot(dat.obs['total_counts'][dat.obs['total_counts']<1000], kde=False, bins=100)\n",
    "plt.show()\n",
    "\n",
    "p5 = sns.displot(dat.obs['total_counts'][dat.obs['total_counts']>1000], kde=False, bins=100)\n",
    "plt.show()\n",
    "\n",
    "#Thresholding decision: genes\n",
    "p6 = sns.displot(dat.obs['n_genes_by_counts'], kde=False, bins=60)\n",
    "plt.show()\n",
    "\n",
    "p7 = sns.displot(dat.obs['n_genes_by_counts'][dat.obs['n_genes_by_counts']<500], kde=False, bins=60)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-calpro]",
   "language": "python",
   "name": "conda-env-.conda-calpro-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
